### ISSUE TITLE
[builtin-terminal-ui] Suspend with ctrl-Z messes up terminal

### ISSUE THREAD

**Issue by ifreund** [mawww/kakoune#3488]
### Steps 

Running the builtin-terminal-ui branch, suspend kakoune with ctrl-Z and then run, for example, `ls -l`.

### Outcome 
![image](https://user-images.githubusercontent.com/12723818/82130361-0407d980-97cb-11ea-92e2-1bd60f8a0ecb.png)

### Expected 
![image](https://user-images.githubusercontent.com/12723818/82130392-4204fd80-97cb-11ea-8c66-a26ed96e1951.png)

**Comment by Screwtapello** [comment]
I can't reproduce this behaviour, using commit 1ce453e and `kak -n`. Can you reproduce with that version and configuration?

**Comment by ifreund** [comment]
Yep, it reproduces on that commit with `kak -n` 
![image](https://user-images.githubusercontent.com/12723818/82140879-53cdbb80-9832-11ea-86da-0448408eea86.png)

**Comment by Screwtapello** [comment]
It looks like Kakoune isn't re-enabling the kernel's newline-to-CRLF translation on suspend. Is the output of `stty` different in the "working" and "broken" cases?

**Comment by ifreund** [comment]
Indeed it is:

Working:
```
speed 38400 baud; line = 0;
-brkint -imaxbel
```

Broken:
```
speed 38400 baud; line = 0;
-brkint -icrnl -imaxbel -opost -isig -iexten noflsh
```

**Comment by Screwtapello** [comment]
Actually, now I look at [the code that handles Ctrl-Z](https://github.com/mawww/kakoune/blob/4a8a1e2af4e794af1f9fe74b24e1ad610e38338e/src/terminal_ui.cc#L391-L415), I see it clears and sets up the mouse handling (`enable_mouse(false)` and `enable_mouse(mouse_enabled)`), it clears and sets up the custom terminal modes (`restore_terminal()` and `setup_terminal()`), it clears and sets up the signal handler, and it even clears and sets up raw mode (`tcsetattr(..., &m_original_termios)` and `set_raw_mode()`)... but both "clear" and "set up" raw mode happen *after* Kakoune suspends itself (`raise(SIGTSTP)`).

Could you try swapping lines 403 and 405 of `terminal_ui.cc` and rebuilding to see if that helps? Specifically, the code that looks like this:

    raise(SIGTSTP); // suspend here

    tcsetattr(STDIN_FILENO, TCSAFLUSH, &m_original_termios);

...should look like this:

    tcsetattr(STDIN_FILENO, TCSAFLUSH, &m_original_termios);

    raise(SIGTSTP); // suspend here

I'm really puzzled why I can't reproduce this, though. I noticed you were using zsh, and I wondered if maybe bash was automatically resetting `stty`, but I couldn't reproduce in zsh either.

**Comment by ifreund** [comment]
That worked! applying this patch as you suggested fixes it
```diff
diff --git a/src/terminal_ui.cc b/src/terminal_ui.cc
index 6429a28d..f36bf3b0 100644
--- a/src/terminal_ui.cc
+++ b/src/terminal_ui.cc
@@ -394,9 +394,10 @@ void TerminalUI::suspend()
     sigaddset(&unblock_sigtstp, SIGTSTP);
     sigprocmask(SIG_UNBLOCK, &unblock_sigtstp, &old_mask);

+    tcsetattr(STDIN_FILENO, TCSAFLUSH, &m_original_termios);
+
     raise(SIGTSTP); // suspend here

-    tcsetattr(STDIN_FILENO, TCSAFLUSH, &m_original_termios);
     set_signal_handler(SIGTSTP, current);
     sigprocmask(SIG_SETMASK, &old_mask, nullptr);
```

**Comment by mawww** [comment]
I believe this is fixed now.

**Comment by ifreund** [comment]
Yes, confirmed fixed by 43a00096b9966e94e800babb8d9c7f1555f63bb3
